### PR TITLE
Added code analyzer

### PR DESCRIPTION
**Pull request by Ambareen09**
## What does this PR do?

Updated the library [utopia-php/messaging](https://github.com/utopia-php/messaging) and implemented a code analyzer. I used [PHPStan](https://phpstan.org/) as the analyzer.

## Test Plan

Screenshot from 2022-10-14 15-54-01

No issues were found upon running the code analyzer!

## Related PRs and Issues

Add Code Analyzer To ‘utopia-php/messaging’(https://github.com/appwrite/appwrite/issues/4381)

### Have you read the [Contributing Guidelines on issues](https://github.com/utopia-php/messaging/blob/master/CONTRIBUTING.md)?

Yes

### CODE REVIEW

**Comment by Ambareen09** [comment]
Please leave a review if any changes are required. Looking forward for a merge!

File: composer.json [hunk 1]
```diff
@@ -34,7 +35,9 @@
   "require-dev": {
     "phpunit/phpunit": "9.5.*",
     "phpmailer/phpmailer": "6.6.*",
-    "squizlabs/php_codesniffer": "^3.6"
+    "squizlabs/php_codesniffer": "^3.6",
+    "vimeo/psalm": "4.0.1",
```
Comment (Meldiron) [inline composer.json:39]: I don't think we need this dependency, we will use PHP Stan for code analysis.

File: composer.json [hunk 2]
```diff
@@ -8,7 +8,8 @@
   "scripts": {
     "test": "vendor/bin/phpunit",
     "lint": "vendor/bin/phpcs",
-    "format": "vendor/bin/phpcbf"
+    "format": "vendor/bin/phpcbf",
+    "check": "./vendor/bin/phpstan analyse --level max src tests"
```
Comment (Meldiron) [inline composer.json:12]: By running `composer check` I got a few errors that needs to be addressed:

CleanShot 2022-10-22 at 18 19 27

**Review by Meldiron — CHANGES_REQUESTED** [review]
Let's please address code analysis errors 🙏 If necessary, we can lower the level of PHP Stan, that should get rid of some errors.

File: composer.json (hunk 2 above)
Comment (Ambareen09) [reply composer.json]: I have fixed errors till level 8. Can you please look into it now?

**Comment by Ambareen09** [comment]
I have ignored some errors, to pass, because I was unable to figure out the correct code. Can you please guide me so as to how to fix them?

**Comment by Ambareen09** [comment]
Screenshot from 2022-10-30 19-01-27

**Comment by Ambareen09** [comment]
@Meldiron  can you please review it?

**Comment by Meldiron** [comment]
Thank you so much for the PR 🤩. We're adding the `hacktoberfest-accepted` label to ensure this PR counts towards your Hacktoberfest contributions count. With that said, please stay active on this PR to address any comments once you receive a review. Happy Hacktoberfest! 🎃

**Comment by stnguyen90** [comment]
@Ambareen09 sorry for the delay. Would you please try and resolve the merge conflicts?

**Comment by Ambareen09** [comment]
I have resolved merge conflict, can you please have a look?

File: src/Utopia/Messaging/Adapter.php [hunk 3]
```diff
@@ -77,11 +76,11 @@ protected function request(
             throw new \Exception('Error:'.\curl_error($ch));
         }
         if (\curl_getinfo($ch, CURLINFO_HTTP_CODE) >= 400) {
-            throw new \Exception($response);
+            throw new \Exception($this->$response);
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapter.php:79]:
```suggestion
            throw new \Exception($response);
```

File: src/Utopia/Messaging/Adapter.php [hunk 4]
```diff
@@ -77,11 +76,11 @@ protected function request(
...
         }
 
         \curl_close($ch);
 
-        return $response;
+        return $this->$response;
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapter.php:84]:
```suggestion
        return $response;
```

File: src/Utopia/Messaging/Adapters/Email.php [hunk 5]
```diff
@@ -21,14 +21,16 @@ public function getMessageType(): string
     /**
      * {@inheritdoc}
      *
+     * @param  EmailMessage  $message Message to send.
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/Email.php:24]:
```suggestion
     * @param  Message $message Message to send.
```

File: src/Utopia/Messaging/Adapters/Email.php [hunk 6]
```diff
@@ -21,14 +21,16 @@ public function getMessageType(): string
...
+     *
      * @throws \Exception
      */
     public function send(Message $message): string
     {
         if (! \is_a($message, $this->getMessageType())) {
             throw new \Exception('Invalid message type.');
         }
-        if (\count($message->getTo()) > $this->getMaxMessagesPerRequest()) {
+        if (\count($message->$this->getTo()) > $this->getMaxMessagesPerRequest()) {
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/Email.php:33]:
```suggestion
        if (\count($message->getTo()) > $this->getMaxMessagesPerRequest()) {
```

File: src/Utopia/Messaging/Adapters/Email/Mock.php [hunk 7]
```diff
@@ -36,14 +38,14 @@ protected function process(Email $message): string
         $mail->SMTPAuth = false;
         $mail->Username = '';
         $mail->Password = '';
-        $mail->SMTPSecure = false;
+        $mail->SMTPSecure = 'false';
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/Email/Mock.php:41]:
```suggestion
        $mail->SMTPSecure = false;
```

File: src/Utopia/Messaging/Adapters/Email/Mock.php [hunk 8]
```diff
@@ -36,14 +38,14 @@ protected function process(Email $message): string
...
         $mail->SMTPAutoTLS = false;
         $mail->CharSet = 'UTF-8';
         $mail->Subject = $message->getSubject();
         $mail->Body = $message->getContent();
         $mail->AltBody = \strip_tags($message->getContent());
-        $mail->setFrom($message->getFrom(), 'Utopia');
-        $mail->addReplyTo($message->getFrom(), 'Utopia');
+        $mail->setFrom($this->$message->getFrom(), 'Utopia');
+        $mail->addReplyTo($this->$message->getFrom(), 'Utopia');
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/Email/Mock.php:48]:
```suggestion
        $mail->setFrom($this->getFrom(), 'Utopia');
        $mail->addReplyTo($this->getFrom(), 'Utopia');
```

File: src/Utopia/Messaging/Adapters/Email/Mock.php [hunk 9]
```diff
@@ -54,6 +56,6 @@ protected function process(Email $message): string
             throw new \Exception($mail->ErrorInfo);
         }
 
-        return true;
+        return 'true';
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/Email/Mock.php:59]:
```suggestion
        return true;
```

File: src/Utopia/Messaging/Adapters/Email/Mock.php [hunk 10]
```diff
@@ -42,8 +44,8 @@ protected function process(Email $message): string
         $mail->Subject = $message->getSubject();
         $mail->Body = $message->getContent();
         $mail->AltBody = \strip_tags($message->getContent());
-        $mail->setFrom($message->getFrom(), 'Utopia');
-        $mail->addReplyTo($message->getFrom(), 'Utopia');
+        $mail->setFrom($this->getFrom(), 'Utopia');
+        $mail->addReplyTo($this->getFrom(), 'Utopia');
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/Email/Mock.php:48]:
```suggestion
        $mail->setFrom($message->getFrom(), 'Utopia');
        $mail->addReplyTo($message->getFrom(), 'Utopia');
```

File: src/Utopia/Messaging/Adapters/Push.php [hunk 11]
```diff
@@ -21,14 +21,16 @@ public function getMessageType(): string
     /**
      * {@inheritdoc}
      *
+     * @param  PushMessage  $message Message to send.
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/Push.php:24]:
```suggestion
     * @param Message $message Message to send.
```

File: src/Utopia/Messaging/Adapters/Push.php [hunk 12]
```diff
@@ -21,14 +21,16 @@ public function getMessageType(): string
...
+     *
      * @throws \Exception
      */
     public function send(Message $message): string
     {
         if (! \is_a($message, $this->getMessageType())) {
             throw new \Exception('Invalid message type.');
         }
-        if (\count($message->getTo()) > $this->getMaxMessagesPerRequest()) {
+        if (\count($message->$this->getTo()) > $this->getMaxMessagesPerRequest()) {
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/Push.php:33]:
```suggestion
        if (\count($message->getTo()) > $this->getMaxMessagesPerRequest()) {
```

File: src/Utopia/Messaging/Adapters/SMS.php [hunk 13]
```diff
@@ -21,14 +21,16 @@ public function getMessageType(): string
     /**
      * {@inheritdoc}
      *
+     * @param  SMSMessage  $message Message to send.
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/SMS.php:24]:
```suggestion
     * @param Message $message Message to send.
```

File: src/Utopia/Messaging/Adapters/SMS.php [hunk 14]
```diff
@@ -21,14 +21,16 @@ public function getMessageType(): string
...
+     *
      * @throws \Exception
      */
     public function send(Message $message): string
     {
         if (! \is_a($message, $this->getMessageType())) {
             throw new \Exception('Invalid message type.');
         }
-        if (\count($message->getTo()) > $this->getMaxMessagesPerRequest()) {
+        if (\count($message->$this->getTo()) > $this->getMaxMessagesPerRequest()) {
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/SMS.php:33]:
```suggestion
        if (\count($message->getTo()) > $this->getMaxMessagesPerRequest()) {
```

File: src/Utopia/Messaging/Adapters/SMS/TextMagic.php [hunk 15]
```diff
@@ -51,7 +51,7 @@ protected function process(SMS $message): string
             ],
             body: [
                 'text' => $message->getContent(),
-                'from' => \ltrim($message->getFrom(), '+'),
+                'from' => \ltrim($this->$message->getFrom(), '+'),
```
Comment (christyjacob4) [inline src/Utopia/Messaging/Adapters/SMS/TextMagic.php:54]:
```suggestion
                'from' => \ltrim($message->getFrom(), '+'),
```

File: tests/e2e/Base.php [hunk 16]
```diff
@@ -6,20 +6,28 @@
...
 
-    protected function getLastEmail(): array
+    /**
+     * @return mixed
+     */
+    protected function getLastEmail(): mixed
     {
         sleep(3);
-
+        // @phpstan-ignore-next-line
```
Comment (christyjacob4) [inline tests/e2e/Base.php:30]:
```suggestion
```

File: tests/e2e/Base.php [hunk 17]
```diff
@@ -6,20 +6,28 @@
...
+    /**
+     * @return mixed
+     *
+     * @throws \Exception If the request fails.
+     */
+    protected function getLastRequest(): mixed
     {
         \sleep(2);
-
+        // @phpstan-ignore-next-line
```
Comment (christyjacob4) [inline tests/e2e/Base.php:17]:
```suggestion
```

**Comment by Ambareen09** [comment]
Hello, I see you have committed a few changes, you can let me know what all modifications do I need to do in order to make this a successful PR

**Comment by abnegate** [comment]
@christyjacob4 Looks like PHPStan checks are failing

**Comment by christyjacob4** [comment]
@Ambareen09 Thanks a lot for your contributions. I'll keep this PR open so it can be resolved at a later point.

Please reach out to me on our [Discord server](https://appwrite.io/discord) if you would like to claim your Appwrite swags! As a way of saying thank you, we would also love to invite you to join the Appwrite organization on GitHub. Please share your GitHub username with us on Discord.

**Comment by Ambareen09** [comment]
> @Ambareen09 Thanks a lot for your contributions. I'll keep this PR open so it can be resolved at a later point.
> 
> Please reach out to me on our [Discord server](https://appwrite.io/discord) if you would like to claim your Appwrite swags! As a way of saying thank you, we would also love to invite you to join the Appwrite organization on GitHub. Please share your GitHub username with us on Discord.

Hey, thanks @christyjacob4 , I have already received my swags (as I opened other PRs too).

**Comment by abnegate** [comment]
Closing as implemented

**Pull request closed without merging**